File: app/src/main/java/com/jefferson/application/br/activity/FolderPickerActivity.java
package com.jefferson.application.br.activity;

import android.content.Intent;
import android.os.Bundle;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.DecelerateInterpolator;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;
import com.jefferson.application.br.FileModel;
import com.jefferson.application.br.R;
import com.jefferson.application.br.adapter.FilePickerAdapter;
import com.jefferson.application.br.app.SimpleDialog;
import com.jefferson.application.br.database.PathsDatabase;
import com.jefferson.application.br.fragment.AlbumFragment;
import com.jefferson.application.br.model.PickerModel;
import com.jefferson.application.br.task.JTask;
import com.jefferson.application.br.util.Storage;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import android.view.Menu;

public class FolderPickerActivity extends MyCompatActivity implements OnItemClickListener {

    private String currentPath;
    private FilePickerAdapter filePickerAdapter;
    private View myOverlay;
    private List<String> paths;
    private int position;
    private FloatingActionButton fab;

    public class MoveFilesTask extends JTask {

        @Override
        public void onException(Exception e) {
            Toast.makeText(FolderPickerActivity.this, "Error!", Toast.LENGTH_SHORT).show();
        }
        ArrayList<String> mMovedArray;
        String folder;
        SimpleDialog dialog;

        public MoveFilesTask(FolderPickerActivity filePicker, String str) {
            this.mMovedArray = new ArrayList<>();
            this.folder = str;
        }

        @Override
        public void onBeingStarted() {
            this.dialog = new SimpleDialog(FolderPickerActivity.this, SimpleDialog.STYLE_PROGRESS);
            this.dialog.setMax(paths.size());
            this.dialog.setTitle(getString(R.string.movendo));
            this.dialog.setProgress(0);
            this.dialog.show();
        }

        @Override
        protected void onUpdated(Object[] objArr) {
            this.dialog.setProgress((Integer) objArr[0]);
        }

        @Override
        public void onFinished() {
            this.dialog.dismiss();
			Intent intent = new Intent();
			intent.putExtra("moved_files", mMovedArray);
            setResult(RESULT_OK, intent);
            finish();
        }

        @Override
        public void workingThread() {

            for (String str : paths) {
                File file = new File(str);
                File newFile = new File(folder, file.getName());
                if (file.renameTo(newFile)) {
					mMovedArray.add(str);
                }
                sendUpdate(dialog.getProgress() + 1);
            }
        }
    }

    @Override
    protected void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.list_view_layout);
        ListView mListView = findViewById(R.id.androidList);
        Toolbar mToolbar = findViewById(R.id.toolbar);
        this.myOverlay = findViewById(R.id.myOverlayLayout);
        this.position = getIntent().getIntExtra("position", -1);
        this.paths = getIntent().getStringArrayListExtra("selection");
        this.currentPath = getIntent().getStringExtra("current_path");
        // applyParentViewPadding(mListView);
        this.filePickerAdapter = new FilePickerAdapter(getModels(this.position), this, position);
        mListView.setAdapter(this.filePickerAdapter);
        mListView.setOnItemClickListener(this);

        setSupportActionBar(mToolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle("Move to");
        }
        fab = findViewById(R.id.fab);
        fab.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View v) {
                    new MoveFilesTask(FolderPickerActivity.this, (filePickerAdapter.models.get(filePickerAdapter.getSelectedItem())).getPath()).start();
                    fab.hide();
                }
            }
        );
        fab.setVisibility(View.GONE);
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long j) {
        int selectedItem = this.filePickerAdapter.getSelectedItem();

        if (selectedItem == -1) {
            fab.setVisibility(View.VISIBLE);
            Animation slideUpAnimation = AnimationUtils.loadAnimation(FolderPickerActivity.this, R.anim.slide_up);
            slideUpAnimation.setInterpolator(new DecelerateInterpolator());
            fab.startAnimation(slideUpAnimation);
        }
        
        if (selectedItem != i) {
            View overlay = view.findViewById(R.id.folder_picker_check_overlay);
            Animation anim = AnimationUtils.loadAnimation(FolderPickerActivity.this, R.anim.fade_in);
            anim.setDuration(250);
            overlay.startAnimation(anim);
            filePickerAdapter.setSelectedItem(i);
        }
    }

    public void update() {
        this.filePickerAdapter.update(getModels(this.position));
        MainActivity mainActivity = MainActivity.getInstance();
        mainActivity.updateFragment(position);
    }

    public void createFolder() {
        View contentView = getLayoutInflater().inflate(R.layout.dialog_edit_text, null);
        final EditText editText = contentView.findViewById(R.id.editTextInput);
        SimpleDialog dialog = new SimpleDialog(this, SimpleDialog.STYLE_INPUT);

        dialog.setContentView(contentView);
        dialog.setTitle(getString(R.string.criar_pasta));
        dialog.setNegativeButton(getString(android.R.string.cancel), null);
        dialog.setPositiveButton(getString(android.R.string.ok), new SimpleDialog.OnDialogClickListener() {

                @Override
                public boolean onClick(SimpleDialog dialog) {
                    String name = editText.getText().toString();
                    String result = AlbumFragment.validateFolderName(name, FolderPickerActivity.this);
                    if (!result.equals(AlbumFragment.FOLDER_NAME_OKAY)) {
                        Toast.makeText(FolderPickerActivity.this, result, Toast.LENGTH_SHORT).show();
                        return false;
                    }
                    boolean success = AlbumFragment.createFolder(FolderPickerActivity.this, name, position) != null;
                    if (success) {
                        update();
                    }
                    return success;
                }
            }
        ).show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_file_picker, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {

        switch (menuItem.getItemId()) {
            case android.R.id.home:   
                finish();
                break;
            case R.id.item_create_folder:
                createFolder();
                break;
        }

        return super.onOptionsItemSelected(menuItem);
    }

    public List<PickerModel> getModels(int i) {
        ArrayList<PickerModel> arrayList = new ArrayList<>();
        File storageAndFolder = Storage.getFolder(i == 0 ? Storage.IMAGE : Storage.VIDEO);
        PathsDatabase.Folder instance = PathsDatabase.Folder.getInstance(this);

        assert storageAndFolder != null;
        File[] listFiles = storageAndFolder.listFiles();
        assert listFiles != null;

        for (File file : listFiles) {

            if (file.isDirectory() && !file.getAbsolutePath().equals(currentPath)) {
                PickerModel pickerModel = new PickerModel();
                File[] dirList = file.listFiles();
                String folderName = instance.getFolderName(file.getName(), position == 0 ?
                        FileModel.IMAGE_TYPE : FileModel.VIDEO_TYPE);

                if (folderName == null) {
                    folderName = file.getName();
                }

                assert dirList != null;
                int length = dirList.length;

                if (length > 0) {
                    pickerModel.setTumbPath(dirList[0].getAbsolutePath());
                }

                pickerModel.setPath(file.getAbsolutePath());
                pickerModel.setName(folderName);
                pickerModel.setSize(length);
                arrayList.add(pickerModel);
            }
        }
        int visibility = arrayList.isEmpty() ? View.VISIBLE: View.GONE;
        if (myOverlay.getVisibility() != visibility) {
            myOverlay.setVisibility(visibility);
        }
        //sort files in alphabetically
        Collections.sort(arrayList, (model1, model2) -> model1.getName().toLowerCase().compareTo(
                model2.getName().toLowerCase()));
        return arrayList;
    }
}

File: app/src/main/java/com/jefferson/application/br/activity/FolderPickerActivity.kt
/*
 * Copyright (C) 2023 Jefferson Calheiros


 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
package com.jefferson.application.br.activity

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.AnimationUtils
import android.view.animation.DecelerateInterpolator
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.jefferson.application.br.model.FileModel
import com.jefferson.application.br.R
import com.jefferson.application.br.activity.MainActivity.Companion.instance
import com.jefferson.application.br.adapter.FilePickerAdapter
import com.jefferson.application.br.app.SimpleDialog
import com.jefferson.application.br.app.SimpleDialog.OnDialogClickListener
import com.jefferson.application.br.database.AlbumDatabase
import com.jefferson.application.br.model.PickerModel
import com.jefferson.application.br.task.JTask
import com.jefferson.application.br.util.AlbumUtils
import com.jefferson.application.br.util.Storage
import java.io.File
import java.util.*

class FolderPickerActivity : MyCompatActivity(), OnItemClickListener {
    private var currentPath: String? = null
    private lateinit var pickerAdapter: FilePickerAdapter
    private var myOverlay: View? = null
    private var paths: List<String>? = null
    private var position = 0
    private lateinit var fab: FloatingActionButton

    inner class MoveFilesTask(str: String) : JTask() {
        override fun onException(e: Exception?) {
            Toast.makeText(this@FolderPickerActivity, "Error!", Toast.LENGTH_SHORT).show()
        }

        private var movedArray: ArrayList<String> = ArrayList()
        private var folder: String
        var dialog: SimpleDialog? = null

        init {
            folder = str
        }

        override fun onStarted() {
            dialog = SimpleDialog(this@FolderPickerActivity, SimpleDialog.STYLE_PROGRESS)
            dialog?.setMax(paths!!.size)
            dialog?.setTitle(getString(R.string.movendo))
            dialog?.setProgress(0)
            dialog?.show()
        }

        override fun onUpdated(objArr:  Array<out Any>?) {
            dialog?.setProgress((objArr?.get(0) as Int))
        }

        override fun onFinished() {
            dialog?.dismiss()
            val intent = Intent()
            intent.putExtra("moved_files", movedArray)
            setResult(RESULT_OK, intent)
            finish()
        }

        override fun workingThread() {
            for (str in paths!!) {
                val file = File(str)
                val newFile = File(folder, file.name)
                if (file.renameTo(newFile)) {
                    movedArray.add(str)
                }
                postUpdate(dialog!!.getProgress() + 1)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.file_picker_layout)
        val listView = findViewById<ListView>(R.id.androidList)
        val mToolbar = findViewById<Toolbar>(R.id.toolbar)
        myOverlay = findViewById(R.id.myOverlayLayout)
        position = intent.getIntExtra("position", -1)
        paths = intent.getStringArrayListExtra("selection")
        currentPath = intent.getStringExtra("current_path")
        pickerAdapter = FilePickerAdapter(getModels(position), this, position)
        listView.adapter = pickerAdapter
        listView.onItemClickListener = this
        setSupportActionBar(mToolbar)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        actionBar?.title = getString(R.string.maove_to)

        fab = findViewById(R.id.fab)
        fab.setOnClickListener {
            MoveFilesTask(
                pickerAdapter.models[pickerAdapter.selectedItem].path!!
            ).start()
            fab.hide()
        }
        fab.visibility = View.GONE
    }

    override fun onItemClick(adapterView: AdapterView<*>?, view: View, i: Int, j: Long) {
        val selectedItem = pickerAdapter.selectedItem
        if (selectedItem == -1) {
            fab.visibility = View.VISIBLE
            val slideUpAnimation =
                AnimationUtils.loadAnimation(this@FolderPickerActivity, R.anim.slide_up)
            slideUpAnimation.interpolator = DecelerateInterpolator()
            fab.startAnimation(slideUpAnimation)
        }
        if (selectedItem != i) {
            val overlay = view.findViewById<View>(R.id.folder_picker_check_overlay)
            val anim = AnimationUtils.loadAnimation(this@FolderPickerActivity, R.anim.fade_in)
            anim.duration = 250
            overlay.startAnimation(anim)
            pickerAdapter.selectedItem = i
        }
    }

    fun updateChanges() {
        val mainActivity = instance
        mainActivity?.updateFragment(position)
        pickerAdapter.update(getModels(position))
    }

    @SuppressLint("InflateParams")
    private fun createFolder() {
        val dialog = SimpleDialog(this, SimpleDialog.STYLE_INPUT)
        dialog.setTitle(getString(R.string.criar_pasta))
        dialog.setNegativeButton(getString(android.R.string.cancel), null)
        dialog.setPositiveButton(getString(android.R.string.ok), object : OnDialogClickListener() {
            override fun onClick(dialog: SimpleDialog): Boolean {
                val name = dialog.getInputText()
                val result = AlbumUtils.validateName(name, this@FolderPickerActivity)
                if (!result.ok) {
                    Toast.makeText(this@FolderPickerActivity, result.message, Toast.LENGTH_SHORT)
                        .show()
                    return false
                }
                val success =
                    AlbumUtils.createAlbum(this@FolderPickerActivity, name, position) != null
                if (success) {
                    updateChanges()
                }
                return success
            }
        }).show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_file_picker, menu)
        return true
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            android.R.id.home -> finish()
            R.id.item_create_folder -> createFolder()
        }
        return super.onOptionsItemSelected(menuItem)
    }

    private fun getModels(i: Int): List<PickerModel> {
        val arrayList = ArrayList<PickerModel>()
        val storageAndFolder = Storage.getFolder(if (i == 0) Storage.IMAGE else Storage.VIDEO, this)
        val instance = AlbumDatabase.getInstance(this)
        val listFiles = storageAndFolder?.listFiles()!!
        for (file in listFiles) {
            if (file.isDirectory && file.absolutePath != currentPath) {
                val pickerModel = PickerModel()
                val dirList = file.listFiles()
                var folderName = instance.getAlbumName(
                    file.name,
                    if (position == 0) FileModel.IMAGE_TYPE else FileModel.VIDEO_TYPE
                )
                if (folderName == null) {
                    folderName = file.name
                }
                val length = dirList!!.size
                if (length > 0) {
                    pickerModel.thumbnailPath = dirList[0].absolutePath
                }
                pickerModel.path = file.absolutePath
                pickerModel.name = folderName
                pickerModel.size = length
                arrayList.add(pickerModel)
            }
        }
        val visibility = if (arrayList.isEmpty()) View.VISIBLE else View.GONE
        if (myOverlay!!.visibility != visibility) {
            myOverlay?.visibility = visibility
        }
        //sort files in alphabetically
        arrayList.sortWith { model1: PickerModel, model2: PickerModel ->
            model1.name!!.lowercase(
                Locale.getDefault()
            ).compareTo(
                model2.name!!.lowercase(Locale.getDefault())
            )
        }
        return arrayList
    }
}